Add background blur feature for supported devices

toger5 opened the pull request:
Fixes https://github.com/element-hq/element-call/issues/121

This adds background blur via @livekit/track-processors for supported browsers.
The option can be found in the video tab of the settings.

The fallback will be a greyed out checkbox (on firefox for instance)
This way firefox users get informed why it does not work on their side if other users tell them where to set it up.

Sadly this is not exactly done the way we would want to set it up see: https://github.com/livekit/track-processors-js/issues/50

- [x] bundle local assets not CDN
- [x] stringify related hang on startup

toger5 committed:
Add background Blur

toger5 committed:
make compatible with unsopported browsers

toger5 commented:
how do we deal with lk modules that dont fullfill the ts lints?

toger5 committed:
add types

toger5 committed:
remove unused dependencies.

toger5 committed:
update blur checkbox (disabled state)

toger5 committed:
i18n

toger5 committed:
eslint + prettier

toger5 committed:
add @types/dom-mediacapture-transform for the remaining type errors

toger5 commented:
>  how do we deal with lk modules that dont fullfill the ts lints?

I found what we needed:
[@types/dom-mediacapture-transform](https://www.npmjs.com/package/@types/dom-mediacapture-transform)

toger5 committed:
properly add @livekit/track-processors

toger5 committed:
dont update if processor already set

toger5 committed:
fixes and gpu based blurring

toger5 committed:
make default false

toger5 committed:
more detailed error message

hughns commented on locales/en-GB/app.json:
```diff
@@ -146,6 +146,9 @@
     },
+    "background_blur_header": "Background",
+    "background_blur_label": "Enable background blurring",
+    "blur_not_supported_by_browser": "(Background blur is not supported by this browser.)",
```
```suggestion
    "blur_not_supported_by_browser": "(Background blur is not supported by this browser)",
```

hughns commented on src/livekit/useLiveKit.ts:
```diff
@@ -9,14 +9,17 @@ import {
 import E2EEWorker from "livekit-client/e2ee-worker?worker";
 import { logger } from "matrix-js-sdk/src/logger";
 import { MatrixRTCSession } from "matrix-js-sdk/src/matrixrtc/MatrixRTCSession";
+import { BackgroundBlur } from "@livekit/track-processors";
```
```suggestion
import { BackgroundBlur as backgroundBlur } from "@livekit/track-processors";
```

hughns commented on src/livekit/useLiveKit.ts:
```diff
@@ -78,13 +83,24 @@ export function useLiveKit(
+  const blur = useMemo(() => {
+    let b = undefined;
+    try {
+      // eslint-disable-next-line new-cap
```
```suggestion
```

hughns commented on src/livekit/useLiveKit.ts:
```diff
@@ -78,13 +83,24 @@ export function useLiveKit(
+    let b = undefined;
+    try {
+      // eslint-disable-next-line new-cap
+      b = BackgroundBlur(15, { delegate: "GPU" });
```
```suggestion
      b = backgroundBlur(15, { delegate: "GPU" });
```

hughns commented on src/livekit/useLiveKit.ts:
```diff
@@ -78,13 +83,24 @@ export function useLiveKit(
       videoCaptureDefaults: {
         ...defaultLiveKitOptions.videoCaptureDefaults,
         deviceId: initialDevices.current.videoInput.selectedId,
+        // eslint-disable-next-line new-cap
```
```suggestion
```

hughns commented on src/room/LobbyView.tsx:
```diff
@@ -16,6 +16,7 @@ import { usePreviewTracks } from "@livekit/components-react";
 import { LocalVideoTrack, Track } from "livekit-client";
 import { useObservable } from "observable-hooks";
 import { map } from "rxjs";
+import { BackgroundBlur } from "@livekit/track-processors";
```
```suggestion
import { BackgroundBlur as backgroundBlur } from "@livekit/track-processors";
```

hughns commented on src/room/LobbyView.tsx:
```diff
@@ -108,6 +111,21 @@ export const LobbyView: FC<Props> = ({
+  const blur = useMemo(() => {
+    let b = undefined;
+    try {
+      // eslint-disable-next-line new-cap
```
```suggestion
```

hughns commented on src/room/LobbyView.tsx:
```diff
@@ -108,6 +111,21 @@ export const LobbyView: FC<Props> = ({
+    let b = undefined;
+    try {
+      // eslint-disable-next-line new-cap
+      b = BackgroundBlur(15, { delegate: "GPU" });
```
```suggestion
      b = backgroundBlur(15, { delegate: "GPU" });
```

hughns commented on src/room/LobbyView.tsx:
```diff
@@ -146,7 +167,23 @@ export const LobbyView: FC<Props> = ({
+    if (!blur) return;
+    const updateBlur = async (showBlur: boolean): Promise<void> => {
+      if (showBlur && !videoTrack?.getProcessor()) {
+        // eslint-disable-next-line new-cap
```
```suggestion
```

hughns commented on src/settings/SettingsModal.tsx:
```diff
@@ -5,10 +5,12 @@ SPDX-License-Identifier: AGPL-3.0-only
 import { MatrixClient } from "matrix-js-sdk/src/matrix";
-import { Root as Form, Text } from "@vector-im/compound-web";
+import { Root as Form, Separator, Text } from "@vector-im/compound-web";
+import { BackgroundBlur } from "@livekit/track-processors";
```
```suggestion
import { BackgroundBlur as backgroundBlur } from "@livekit/track-processors";
```

hughns commented on src/settings/SettingsModal.tsx:
```diff
@@ -70,6 +73,42 @@ export const SettingsModal: FC<Props> = ({
+    let canBlur = true;
+    try {
+      // eslint-disable-next-line new-cap
+      BackgroundBlur(15);
```
```suggestion
      backgroundBlur(15);
```

hughns commented on src/settings/SettingsModal.tsx:
```diff
@@ -70,6 +73,42 @@ export const SettingsModal: FC<Props> = ({
+    const [blur, setBlur] = useSetting(backgroundBlurSetting);
+    let canBlur = true;
+    try {
+      // eslint-disable-next-line new-cap
```
```suggestion
```

hughns commented on src/room/LobbyView.tsx:
```diff
@@ -108,6 +111,21 @@ export const LobbyView: FC<Props> = ({
       muteStates.audio.enabled && { deviceId: devices.audioInput.selectedId },
   );
 
+  // eslint-disable-next-line new-cap
```
```suggestion
```

hughns commented on locales/en-GB/app.json:
```diff
@@ -146,6 +146,9 @@
       "effect_volume_label": "Sound effect volume"
     },
+    "background_blur_header": "Background",
+    "background_blur_label": "Enable background blurring",
```
```suggestion
    "background_blur_label": "Blur the background of the video",
```

hughns reviewed (dismissed):
I've suggested some cleanup. Furthermore I think you could deduplicate some of the code into something like a `useBlur()` or `canBlur()`. This might make adding tests easier too.

toger5 committed:
review

toger5 committed:
review

toger5 commented:
> Furthermore I think you could deduplicate some of the code into something like a useBlur() or canBlur(). This might make adding tests easier too.

I think @robintown has an idea how we might be able to reuse the track from the lobby to the in call view. Than this might come for free.

But in this pr we should at least get rid of the duplicated `let blur = useMemo` sections

toger5 committed:
disable blur in local track again

hughns commented on locales/en-GB/app.json:
```diff
@@ -150,6 +150,9 @@
     },
+    "background_blur_header": "Background",
+    "background_blur_label": "Blur the background of the video",
+    "blur_not_supported_by_browser": "(Background blur is not supported by this browser)",
```
```suggestion
    "blur_not_supported_by_browser": "(Background blur is not supported by this device)",
```

...because this message also appears when EC is embedded as widget and the user isn't aware that it is a browser.

hughns commented:
Currently this PR causes some (large) assets to be loaded from jsdelivr.net and googleapis.com at runtime:

![image](https://github.com/user-attachments/assets/7beaabf8-d392-4bb6-9cba-6ae6be02424b)

hughns commented:
Marking draft until the assets are loaded from local EC.

toger5 committed:
refactor
 - still should not work without a fixing upstream LK:
https://github.com/livekit/components-js/pull/1042
https://github.com/livekit/components-js/pull/1043

hughns committed:
Use local assets for blurring

hughns committed:
Lint

hughns committed:
Wording change

hughns committed:
Lint

sandhose commented on src/livekit/BlurBackgroundTransformer.ts:
```diff
@@ -0,0 +1,62 @@
+// n.b. this only includes the SIMD versions of the WASM files which have good support:
+// https://caniuse.com/?search=simd
+const wasmFileset: WasmFileset = {
+  wasmLoaderPath: new URL(
+    "../../node_modules/@mediapipe/tasks-vision/wasm/vision_wasm_internal.js",
```
I think it would be better to use

```js
import wasmLoaderPath from "@mediapipe/tasks-vision/wasm/vision_wasm_internal.js?url";
```

hughns replied:
I did try that, but I ended up with a vite error and so did this approach instead. 🤷

hughns committed:
Merge branch 'livekit' into toger5/track-processor-blur

hughns committed:
.

hughns committed:
Format

toger5 committed:
Merge branch 'livekit' into toger5/track-processor-blur

toger5 committed:
bump components-js (not yet working)

hughns committed:
Merge branch 'livekit' into toger5/track-processor-blur

hughns committed:
Lint

toger5 committed:
Revert "Bump matrix-js-sdk for bundle improvements (#2911)"

This reverts commit 6b8dddfaaad68fd0bd1f4f5c0c2fb7b351775ffe.

toger5 committed:
working version

hughns committed:
Merge branch 'livekit' into toger5/track-processor-blur

hughns committed:
Exclude tasks-vision from knip

hughns committed:
Try reenabling local implementation of BlurBackgroundTransformer

hughns commented:
Marking as draft whilst we fix these:

> hook.js:608 error when trying to pipe IndexSizeError: Failed to construct 'ImageData': The source width is zero or not a number.
    at Vz (index.mjs:291:1)
    at MZe.drawVirtualBackground (index.mjs:230:14)
    at MZe.transform (index.mjs:205:21)
    at Object.transform (index.mjs:117:39)

It was also showing a green colour instead of blurring at one point. I'm not able to reliably reproduce that.

hughns committed:
Merge branch 'livekit' into toger5/track-processor-blur

hughns commented:
Also this:

![image](https://github.com/user-attachments/assets/6a67a182-69dd-47c5-a317-e5df45ee917b)

toger5 commented:
I also was not yet able to fix the following issues:
  -  configure to not use blur -> press on a recent call from the list. -> turn on the blur it does not activate blur for me error when trying to pipe IndexSizeError: Failed to construct 'ImageData': The source width is zero or not a number.
  -  unrelated but I now remember that I looked into this as well: start a call -> wait for the video in the lobby -> open settings observe that the video feed turns black for a couple of seconds.

See: https://matrix.to/#/!dXHEZfLxIQzAELxzQd:matrix.org/$i8KmqSqyJN4s65D8oXTjcSLTVgFo2uB30HWVlWAVk_Q?via=matrix.org&via=element.io&via=robin.town
and: https://matrix.to/#/!dXHEZfLxIQzAELxzQd:matrix.org/$ad9iclLVvo8doof3p8iFQBkL43koY-YxbV4PcCWUQfo?via=matrix.org&via=element.io&via=robin.town
for demo videos about those issues.

hughns committed:
Merge branch 'livekit' into toger5/track-processor-blur

hughns committed:
Fix @matrix-org/matrix-sdk-crypto-wasm version

hughns committed:
Bump @livekit/track-processors to ignore empty frames

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/element-hq/element-call?pullRequest=2812) <br/>All committers have signed the CLA.

toger5 committed:
Merge branch 'livekit' into toger5/track-processor-blur

toger5 committed:
package and lock file

robintown committed:
Merge branch 'livekit' into toger5/track-processor-blur

robintown committed:
Upgrade LiveKit track processors

robintown committed:
Dedupe Mediapipe dependencies

robintown committed:
Get blurring working again

robintown committed:
Cleanup + use a better check for blurring support

robintown commented:
I believe this is working again. Also: the LiveKit team added support for Firefox and Safari in the meantime via a slightly lower performance API. :tada: Still have tests and a couple of personal nits to fix before I ask for another review.

robintown committed:
Merge branch 'livekit' into toger5/track-processor-blur

robintown committed:
Remove unneeded Knip override for MediaPipe

robintown committed:
Import the segmenter model in a more statically-analyzable way

robintown committed:
Document the background transformer a bit better

robintown committed:
Make background blurring actually react to track restarts

robintown committed:
Don't lie about the type of the video track

robintown committed:
Fix tests

robintown committed:
Upgrade TypeScript for ImageBufferSource support

robintown commented:
The type error seems to be due to a bug in the @types/dom-webcodecs type declarations. I've opened [a PR to fix it](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/72625).

robintown committed:
Inline a patched version of WebCodecs types

robintown commented:
Since we want to get this change published in an RC pretty soon, I figured out a way to unblock it: the patched versions of the type declarations are now inlined into our own source tree.

robintown committed:
Clean up a string

robintown committed:
Merge branch 'livekit' into toger5/track-processor-blur

BillCarsonFr commented on locales/en/app.json:
```diff
@@ -164,6 +164,9 @@
       "effect_volume_description": "Adjust the volume at which reactions and hand raised effects play.",
       "effect_volume_label": "Sound effect volume"
     },
+    "background_blur_header": "Background",
```
Are these keys added in localazy? I couldn't find them

robintown replied:
They will be automatically added to Localazy as soon as this PR is merged

BillCarsonFr commented on src/@types/dom-mediacapture-transform.d.ts:
```diff
@@ -0,0 +1,137 @@
+/* eslint-disable */
```
I wonder what copyright should be there

robintown replied:
Good point, I've copied over the MIT license notice

BillCarsonFr reviewed (approved):
Just some small questions. 
Also there are no tests :/

robintown committed:
Add license notices to inlined type declarations

robintown commented:
> Also there are no tests :/

The most valuable thing to test here would be the part that makes sure the processor is always applied to the user's video track across track restarts, but this is also the most difficult file in EC to add tests for currently due to the complex React hook dynamics and interaction with LiveKit. I think this needs some real strategy if we want to start testing useLiveKit.ts.
